deltachat-ios/ViewModel/ContactDetailViewModel.swift

```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+protocol ContactDetailViewModelProtocol {
+    var contactId: Int { get }
+    var contact: DcContact { get }
+    var numberOfSections: Int { get }
+    func numberOfRowsInSection(_ : Int) -> Int
+    func typeFor(section: Int) -> ContactDetailViewModel.SectionType
+    func update(sharedChatCell: ContactCell, row index: Int)
+    func getSharedChatIdAt(indexPath: IndexPath) -> Int
+    func titleFor(section: Int) -> String?
+}
+
+class ContactDetailViewModel: ContactDetailViewModelProtocol {
+
+    let context: DcContext
+    enum SectionType {
+        case startChat
+        case sharedChats
+        case blockContact
+    }
+
+    var contactId: Int
+
+    var contact: DcContact
+    private let sharedChats: DcChatlist
+    private let startChatOption: Bool
+
+    private var sections: [SectionType] = []
+
+    init(contactId: Int, startChatOption: Bool, context: DcContext) {
+        self.context = context
+        self.contactId = contactId
+        self.contact = DcContact(id: contactId)
+        self.startChatOption = startChatOption
+        self.sharedChats = context.getChatlist(flags: 0, queryString: nil, queryId: contactId)
+
+        if startChatOption {
+            sections.append(.startChat)
+        }
+        if sharedChats.length > 0 {
+            sections.append(.sharedChats)
+        }
+        sections.append(.blockContact)
+    }
+
+    func typeFor(section: Int) -> ContactDetailViewModel.SectionType {
+        return sections[section]
+    }
+
+    var numberOfSections: Int {
+        return sections.count
+    }
+
+    func numberOfRowsInSection(_ section: Int) -> Int {
+        switch sections[section] {
+        case .sharedChats: return sharedChats.length
+        case .blockContact, .startChat: return 1
+        }
+    }
+
+    func getSharedChatIdAt(indexPath: IndexPath) -> Int {
+        let index = indexPath.row
+        // assert(sections[indexPath.section] == .sharedChats)
+        return sharedChats.getChatId(index: index)
+    }
+
+    func update(sharedChatCell cell: ContactCell, row index: Int) {
+        let chatId = sharedChats.getChatId(index: index)
+        let summary = sharedChats.getSummary(index: index)
+        let unreadMessages = context.getUnreadMessages(chatId: chatId)
```

Comment:
while this is probably fine for the "shared chats" view, however, this may make problems when used at the "normal" chatlist. imu, these two functions are called for all chats in the chatlist - even if they are not in view and never will be in view - assume a list with 500 chats, this is a lot of overhead.

again, for 20 or so shared chats, this is not an issue, however, maybe this is already a good moment to think this over and use eg. a function that loads the data as really needed - the recycler view already works this way afaik, so that should be easy.

Reply:
this method is called as part of the cell-recycling process. it will be called for each cell that appears on screen right before it is scrolled in. The length of the chatList has no effect on how often this method is called.

Reply:
i see, that would be just fine :)

i probably mixed that up with the loop (is it a loop?) `let _ = (0..<chatList.length).map` from https://github.com/deltachat/deltachat-ios/pull/500/files#diff-5e38f3d38776c60d414d5a59520ba530R114 - that would be bad then?